File: web/src/lib/studyCalendar.ts
import type { Study } from '../api/personal';

export function studyMonth(today: string, offset = 0) {
  // Календарь следует дате серии с сервера; часы и часовой пояс браузера
  // не должны переносить занятие на соседний день.
  const parsed = new Date(`${today}T12:00:00Z`);
  if (!Number.isFinite(parsed.getTime())) return null;
  const first = new Date(Date.UTC(parsed.getUTCFullYear(), parsed.getUTCMonth() + offset, 1));
  const count = new Date(Date.UTC(first.getUTCFullYear(), first.getUTCMonth() + 1, 0)).getUTCDate();
  return {
    key: first.toISOString().slice(0, 7),
    label: first.toLocaleDateString('ru-RU', { month: 'long', year: 'numeric', timeZone: 'UTC' }),
    padding: (first.getUTCDay() + 6) % 7,
    days: Array.from({ length: count }, (_, i) => `${first.toISOString().slice(0, 7)}-${String(i + 1).padStart(2, '0')}`),
  };
}

export function latestStudy(initial?: Study, live?: Study | null): Study | undefined {
  if (!initial) return live ?? undefined;
  if (!live) return initial;
  // Свежая серверная статистика важнее старого локального снимка.
  return (Date.parse(live.asOf ?? '') || 0) >= (Date.parse(initial.asOf ?? '') || 0) ? live : initial;
}

File: frontend/lib/utils/study_calendar.dart
/// Календарь использует дату серии с сервера, не часовой пояс телефона.
class StudyMonth {
  StudyMonth(this.first);
  final DateTime first;
  int get padding => first.weekday - 1;
  int get dayCount => DateTime.utc(first.year, first.month + 1, 0).day;
  String date(int day) =>
      '${first.year.toString().padLeft(4, '0')}-${first.month.toString().padLeft(2, '0')}-${day.toString().padLeft(2, '0')}';
  String get label => '${const [
        'Январь',
        'Февраль',
        'Март',
        'Апрель',
        'Май',
        'Июнь',
        'Июль',
        'Август',
        'Сентябрь',
        'Октябрь',
        'Ноябрь',
        'Декабрь'
      ][first.month - 1]} ${first.year}';
}

StudyMonth? studyMonth(String today, [int offset = 0]) {
  if (!RegExp(r'^\d{4}-\d{2}-\d{2}$').hasMatch(today)) return null;
  final parsed = DateTime.tryParse('${today}T12:00:00Z');
  if (parsed == null || parsed.toIso8601String().substring(0, 10) != today) {
    return null;
  }
  return StudyMonth(DateTime.utc(parsed.year, parsed.month + offset));
}

Map<String, dynamic>? latestStudyData(
    Map<String, dynamic>? initial, Map<String, dynamic>? live) {
  if (initial == null) return live;
  if (live == null) return initial;
  int stamp(Map<String, dynamic> value) =>
      DateTime.tryParse('${value['asOf'] ?? ''}')?.millisecondsSinceEpoch ?? 0;
  return stamp(live) >= stamp(initial) ? live : initial;
}
